README.md
# Hassio Addons
[![Build Status](https://travis-ci.org/editter/hassio-addons.svg?branch=master)](https://travis-ci.org/editter/hassio-addons)
[![Donate](https://img.shields.io/badge/Donate-PayPal-blue.svg)](https://paypal.me/editter)
[![Patreon](https://img.shields.io/badge/Donate-Patreon-blue.svg)](https://www.patreon.com/editter)

## [Life360](https://github.com/editter/hassio-addons/tree/master/life360)

[![t](https://images.microbadger.com/badges/version/editter/life360-amd64.svg)](https://microbadger.com/images/editter/life360-amd64 "Get your own version badge on microbadger.com")
[![](https://images.microbadger.com/badges/image/editter/life360-amd64.svg)](https://microbadger.com/images/editter/life360-amd64 "Get your own image badge on microbadger.com")

Presence detection using [Life360](https://life360.com) location service


life360/CHANGELOG.md
## Releases

### 0.0.13

- Allowed refresh seconds to be a float instead of an int
- Updated base image to fix API bugs

### 0.0.12

- Added hassio api key

### 0.0.11

- Removed the server if there isn't a webhook

### 0.0.10

- Removed the requirement for SSL certificates.  Not webhook is setup in that case
- Separated readme.md into two files so it isn't so ugly in hassio web page

### 0.0.9

- Moved attributes for HTTP so they are accessible in sensor

### 0.0.8

- Added is_charging option per user request
- Removed address from MQTT message

### 0.0.7

- Checked user location data to ensure it exists (in some cases it doesn't seem to get populated)
- Added initial information on startup to help with debugging

### 0.0.6

- Added the option to use HTTP for reporting the location instead of forcing MQTT. If you want to continue using MQTT there is no change required for you, just providing options for people.

### 0.0.5

- Added refresh_minutes setting for quicker refresh rates
- Removed file logging (not sure if they are viewable in Hass.io)
- Minor bug fixes

### 0.0.4

- Made preface the start of the MQTT topic and removed the slash in code
- Added some minor security to the webhook
- Removed the web interface due to it not being secure (may add it back in when I figure out how to make it secure)


life360/server.js
/*eslint quotes: [2, single"]*/
/*eslint-env es6*/
/*jslint esversion: 6 */

// 'use strict';

const winston = require('winston'),
  express = require('express'),
  https = require('https'),
  expressWinston = require('express-winston'),
  bodyparser = require('body-parser'),
  mqtt = require('mqtt'),
  async = require('async'),
  path = require('path'),
  jsonfile = require('jsonfile'),
  fs = require('fs'),
  Axios = require('axios');

const baseAuthKey = 'cFJFcXVnYWJSZXRyZTRFc3RldGhlcnVmcmVQdW1hbUV4dWNyRUh1YzptM2ZydXBSZXRSZXN3ZXJFQ2hBUHJFOTZxYWtFZHI0Vg==';

const CONFIG_DIR = process.env.CONFIG_DIR || './data',
  STATE_FILE = path.join(CONFIG_DIR, 'state.json'),
  EVENTS_LOG = path.join(CONFIG_DIR, 'events.log'),
  ACCESS_LOG = path.join(CONFIG_DIR, "access.log"),
  ERROR_LOG = path.join(CONFIG_DIR, 'error.log'),
  OPTIONS = path.join(CONFIG_DIR, 'options.json'),
  CURRENT_VERSION = jsonfile.readFileSync(path.join('package.json')).version,
  PORT = 8081;


let config = jsonfile.readFileSync(OPTIONS),
  state = loadSavedState({
    savedToken: null,
    circles: [
      /*{
            id: '',
            name: null,
          }*/
    ],
    queueHistory: {},
    members: [],
    places: [],
    CURRENT_VERSION
  });

// write all events to disk as well
winston.add(winston.transports.File, {
  filename: EVENTS_LOG,
  json: false
});
const url = config.mqtt_host.includes('://') ? config.mqtt_host : `mqtt://${config.mqtt_host}`;

const app = express(),
  isProd = process.env.IS_PROD === 'true',
  client = mqtt.connect(`${url}:${config.mqtt_port}`, {
    username: config.mqtt_username,
    password: config.mqtt_password
  });


function loadSavedState(defaults) {
  let output;
  try {
    output = jsonfile.readFileSync(STATE_FILE);
  } catch (ex) {
    winston.info("No previous state found, continuing.");

    output = defaults;
  }
  return output;
}

async function refreshToken() {

  // try {
  const headers = {
    headers: {
      Authorization: "basic " + baseAuthKey
    }
  };
  const authResponse = await Axios.post("https://api.life360.com/v3/oauth2/token.json", {
    grant_type: "password",
    username: config.life360_user,
    password: config.life360_password
  }, headers);

  if (authResponse.data.access_token) {
    state.savedToken = authResponse.data.token_type + " " + authResponse.data.access_token;
  } else {
    throw Error("No token was returned. " + authResponse.data.errorMessage)
  }

  headers.headers = {
    Authorization: state.savedToken
  };
  const circleResponse = await Axios.get("https://api.life360.com/v3/circles.json", headers);
  if (Array.isArray(circleResponse.data.circles)) {
    state.circles = [];
    circleResponse.data.circles.forEach(circ => {
      state.circles.push({
        id: circ.id,
        name: circ.name
      });

    });

  } else {
    winston.warn("No circles were returned from the request. " + circleResponse.data.errorMessage);
  }
  state.places = [];
  state.circles.forEach(async(circ) => {

    const placesResponse = await Axios.get(`https://api.life360.com/v3/circles/${circ.id}/places.json`, headers);
    if (Array.isArray(placesResponse.data.places)) {
      placesResponse.data.places.forEach(x => state.places.push(x));
    }

    if (config.host_url && config.host_url.length > 0) {
      await Axios.delete(`https://api.life360.com/v3/circles/${circ.id}/webhook.json`, headers);
      const returnUrl = config.host_url + "/webhook";
      const webhookResponse = await Axios.post(`https://api.life360.com/v3/circles/${circ.id}/webhook.json`, {
        url: returnUrl
      }, headers);

      if (!webhookResponse.data.hookUrl) {
        winston.warn("Webhook was not created. " + circleResponse.data.errorMessage);
      }
    }
  });

}

function formatLocation(input) {
  //  {"longitude": 1.0,"gps_accuracy": 60,"latitude": 2.0,"battery_level": 99.9}
  return {
    longitude: parseFloat(input.location.longitude),
    latitude: parseFloat(input.location.latitude),
    gps_accuracy: parseInt(input.location.accuracy),
    battery_level: parseInt(input.location.battery),
    is_intransit: parseInt(input.location.inTransit),
    short_address: input.location.shortAddress,
    // id: input.id,
    name: input.firstName
  }
}

async function refreshState() {
  const topic = `/${config.preface}/location`;

  try {
    state.members = await getData();
    const queueData = state.members.map(x => formatLocation(x));
    state.queueHistory[topic] = queueData;
    jsonfile.writeFileSync(STATE_FILE, state, {
      spaces: 2
    });

    queueData.forEach(msg => {
      const userTopic = `${topic}/${msg.name}`;
      winston.info('Message was written to ' + userTopic);
      client.publish(userTopic, JSON.stringify(msg), {
        retain: true
      });
    });

  } catch (err) {
    winston.error(err);
  }
}

async function getData() {
  let results = [];
  if (!state.savedToken) {
    await refreshToken();
  }

  const headers = {
    headers: {
      Authorization: state.savedToken || 'un-auth-request'
    }
  };

  return new Promise((accept, reject) => {
    async.forEach(state.circles, (circ, callback) => {
      Axios.get(`https://api.life360.com/v3/circles/${circ.id}/members.json`, headers)
        .then(membersResponse => {
          membersResponse.data.members.forEach(x => results.push(x));
          callback();
        })
        .catch(err => {
          if (err.status === 401) {
            state.savedToken = null;
          }
          callback(err);
        });
    }, (err) => {
      if (err) {
        reject(err);

      } else {
        accept(results);

      }
    });
  });
}

// main flow
async.series([
    function init(next) {
      refreshToken().then(() => {
        winston.info('init/refresh token successful');
        next();

      }).catch(err => {
        winston.error('Error getting initial token request. Ending process');
        winston.error(err);
        process.exit(1);

      });

    },
    function configurePolling(next) {
      winston.info("Configuring auto update");

      // save current state every 15 minutes
      setInterval(refreshState, 15 * 60 * 1000);
      refreshState();
      process.nextTick(next);
    },
    function setupApp(next) {
      winston.info("Configuring server");

      // accept JSON
      app.use(bodyparser.json());

      // log all requests to disk
      app.use(expressWinston.logger({
        transports: [
          new winston.transports.File({
            filename: ACCESS_LOG,
            json: false
          })
        ]
      }));

      // webhook event from life360
      app.all("/webhook", (req, res) => {
        winston.info('webhook call received');
        refreshState();
      });

      app.all("/get-recent", (req, res) => {
        const data = req.body;
        winston.info('Incoming request for location');
        const recent = state.members.map(x => formatLocation(x)).filter(x => x.name === data.name);
        let response = '';
        if (recent.length > 0) {
          // respond with name is at recent[0].short_address
          const person = recent[0];
          response = `${data.name} is ${person.is_inTransit===0 ? 'moving near':'at'} ${person.short_address}`;
        } else {
          // respond with I couldnt find name
          response = `I could not find ${data.name}`
        }

        res.send(response).end();
      });

      app.get('/check', (req, res, next) => {
        let output = "Life 360 Plugin";
        output += "<h3>Places</h3><br />"
        output += "<pre><code>";
        output += JSON.stringify(state.places.map(x => new Object({
          name: x.name,
          latitude: x.latitude,
          longitude: x.longitude,
          radius: x.radius
        })), null, 4);
        output += "</pre></code>";
        output += "<h3>Circles</h3><br />"
        output += "<pre><code>";
        output += JSON.stringify(state.circles.map(x => new Object({
          name: x.name
        })), null, 4);
        output += "</pre></code>";
        output += "<h3>Circles</h3><br />"
        output += "<pre><code>";
        output += JSON.stringify(state.members.map(x => formatLocation(x)), null, 4);
        output += "</pre></code>";
        res.send(output).end();
      });

      // log all errors to disk
      app.use(expressWinston.errorLogger({
        transports: [
          new winston.transports.File({
            filename: ERROR_LOG,
            json: false
          })
        ]
      }));

      // proper error messages with Joi
      app.use((err, req, res, next) => {
        winston.error(err);

        res.status(500).send({
          errorMessage: 'An unhandled error occured'
        });
      });

      if (isProd && (config.cert_file.length === 0 && config.key_file.length === 0)) {
        winston.error('SSL Certs are required because this addon receives data from a third party so we want their data safely');
        process.exit(1);
      } else if (!isProd) {
        winston.info('NO certificate files found, listing via http');
        app.listen(PORT, next);
      } else {
        winston.info('Listing via https');
        const options = {
          cert: fs.readFileSync(path.join('/ssl/', config.cert_file)),
          key: fs.readFileSync(path.join('/ssl/', config.key_file))
        };
        https.createServer(options, app).listen(PORT, next);

      }

    }
  ],
  (error) => {
    if (error) {
      return winston.error(error);
    }
    winston.info("Listening at port %s", PORT);
  });

life360/server1.ts
/**
 *
 */

import { Member, Place, Circle } from './life360';

import * as winston from 'winston';
import * as express from 'express';
import * as https from 'https';
const expressWinston: any = require('express-winston');
import * as bodyparser from 'body-parser';
import * as mqtt from 'mqtt';
import * as async from 'async';
import * as jsonfile from 'jsonfile';
import * as fs from 'fs';
import * as path from 'path';
import * as Axios from 'axios';
import { Request, NextFunction, json } from 'express';
import { Response } from 'express-serve-static-core';
import * as guid from 'guid';
import { app as AlexaApp, request as AlexaRequest, response as AlexaResponse } from 'alexa-app';

const baseAuthKey = 'cFJFcXVnYWJSZXRyZTRFc3RldGhlcnVmcmVQdW1hbUV4dWNyRUh1YzptM2ZydXBSZXRSZXN3ZXJFQ2hBUHJFOTZxYWtFZHI0Vg==';

const CONFIG_DIR = process.env.CONFIG_DIR || './data',
  STATE_FILE = path.join(CONFIG_DIR, 'state.json'),
  OPTIONS_FILE = path.join(CONFIG_DIR, 'options.json'),
  EVENTS_LOG = path.join(CONFIG_DIR, 'events.log'),
  ACCESS_LOG = path.join(CONFIG_DIR, "access.log"),
  ERROR_LOG = path.join(CONFIG_DIR, 'error.log'),
  CURRENT_VERSION: string = jsonfile.readFileSync(path.join('package.json')).version,
  isProd = process.env.IS_PROD === 'true',
  app = express(),
  alexaApp = new AlexaApp('alexa'),
  PORT = 8081;


let config = jsonfile.readFileSync(OPTIONS_FILE) as {
  mqtt_host: string;
  mqtt_port: number;
  mqtt_username: string;
  mqtt_password: string;
  preface: string;
  cert_file: string;
  key_file: string;
  host_url: string;
  life360_user: string;
  life360_password: string;
},
  state = loadSavedState({
    saved_token: null as string | null,
    circles: [] as { id: string, name: string }[],
    queue_history: {} as { [prop: string]: any },
    members: [] as Member[],
    places: [] as Place[],
    access_token: guid.raw(),
    CURRENT_VERSION
  });

// write all events to disk as well
winston.add(winston.transports.File, {
  filename: EVENTS_LOG,
  json: false
});
const url = config.mqtt_host.includes('://') ? config.mqtt_host : `mqtt://${config.mqtt_host}`,
  client = mqtt.connect(`${url}`, {
    username: config.mqtt_username,
    password: config.mqtt_password,
    port: config.mqtt_port,
  });


function loadSavedState<T>(defaults: T): T {
  try {
    return jsonfile.readFileSync(STATE_FILE);
  } catch (ex) {
    winston.info("No previous state found, setting to defaults.");
    return defaults;
  }

}

async function refreshToken() {

  const headers = {
    headers: {
      Authorization: "basic " + baseAuthKey
    }
  };
  const authResponse = await Axios.post<{ access_token: string, token_type: string, errorMessage?: string }>("https://api.life360.com/v3/oauth2/token.json", {
    grant_type: "password",
    username: config.life360_user,
    password: config.life360_password
  }, headers);

  if (authResponse.data.access_token) {
    state.saved_token = authResponse.data.token_type + " " + authResponse.data.access_token;
  } else {
    throw Error("No token was returned. " + authResponse.data.errorMessage)
  }

  headers.headers = {
    Authorization: state.saved_token
  };
  const circleResponse = await Axios.get<{ circles: Circle[], errorMessage?: string }>("https://api.life360.com/v3/circles.json", headers);
  if (Array.isArray(circleResponse.data.circles)) {
    state.circles = [];
    circleResponse.data.circles.forEach(circ => {
      state.circles.push({
        id: circ.id,
        name: circ.name
      });

    });

  } else {
    winston.warn("No circles were returned from the request. " + circleResponse.data.errorMessage);
  }

  return new Promise<void>((accept, reject) => {
    state.places = [];

    async.forEach(state.circles, async (circ, callback) => {
      const placesResponse = await Axios.get<{ places: Place[], errorMessage: string }>(`https://api.life360.com/v3/circles/${circ.id}/places.json`, headers);
      if (Array.isArray(placesResponse.data.places)) {
        placesResponse.data.places.forEach(x => state.places.push(x));
      }

      if (config.host_url && config.host_url.length > 0) {
        await Axios.delete(`https://api.life360.com/v3/circles/${circ.id}/webhook.json`, headers);
        const returnUrl = config.host_url + "/webhook?access_token=" + state.access_token;
        const webhookResponse = await Axios.post(`https://api.life360.com/v3/circles/${circ.id}/webhook.json`, {
          url: returnUrl
        }, headers);

        if (!webhookResponse.data.hookUrl) {
          winston.warn("Webhook was not created. " + circleResponse.data.errorMessage);
        }
      }
      callback();
    }, (err) => {
      if (err) {
        reject(err);

      } else {
        accept();

      }
    });

  });

}


function formatLocation(input: Member) {
  return {
    longitude: parseFloat(input.location.longitude),
    latitude: parseFloat(input.location.latitude),
    gps_accuracy: parseInt(input.location.accuracy),
    battery_level: parseInt(input.location.battery),
    is_intransit: parseInt(input.location.inTransit),
    address1: input.location.address1,
    name: input.firstName
  }
}

async function refreshState() {
  const topic = `/${config.preface}/location`;

  try {
    state.members = await getData();
    const queueData = state.members.map(x => formatLocation(x));
    state.queue_history[topic] = queueData;
    jsonfile.writeFileSync(STATE_FILE, state, {
      spaces: 2
    });

    queueData.forEach(msg => {
      const userTopic = `${topic}/${msg.name}`;
      winston.info('Message was written to ' + userTopic);
      client.publish(userTopic, JSON.stringify(msg), { qos: 1, retain: true });
    });

  } catch (err) {
    winston.error(err);
  }
}

async function getData() {
  if (!state.saved_token) {
    await refreshToken();
  }

  return new Promise<Member[]>((accept, reject) => {
    const headers = {
      headers: {
        Authorization: state.saved_token || 'un-auth-request'
      }
    };

    let results: Member[] = [];
    async.forEach(state.circles, (circ, callback) => {
      Axios.get<{ members: Member[] }>(`https://api.life360.com/v3/circles/${circ.id}/members.json`, headers)
        .then(membersResponse => {
          membersResponse.data.members.forEach(x => results.push(x));
          callback();
        })
        .catch(err => {
          if (err.status === 401) {
            state.saved_token = null;
          }
          callback(err);
        });
    }, (err) => {
      if (err) {
        reject(err);

      } else {
        accept(results);

      }
    });
  });
}

// main flow
async.series([
  function init(next) {
    refreshToken().then(() => {
      winston.info('init/refresh token successful');
      next();

    }).catch(err => {
      winston.error('Error getting initial token request. Ending process');
      next(err)

    });

  },
  function configurePolling(next) {
    winston.info("Configuring auto update");

    // save current state every 15 minutes
    setInterval(refreshState, 15 * 60 * 1000);
    refreshState();
    process.nextTick(next);
  },
  function alexaSkill(next) {
    winston.info("Configuring alexa skill");
    alexaApp.express({ expressApp: app, debug: true });
    alexaApp.pre = (request, response, type) => {
      winston.info(JSON.stringify(request.getSession().details));
    }

    alexaApp.intent('GetPosition', {
      slots: { "NAME": "AMAZON.US_FIRST_NAME" },
      utterances: [
        "where is {NAME}",
        "where is {NAME} right now",
        "what is the location of {NAME}",
      ]
    }, (request, response) => {
      const slotName = request.slot('NAME')
      winston.info(`Incoming request for alexa location name='${name}'`);
      refreshState();
      const recent = state.members.map(x => formatLocation(x)).filter(x => x.name === slotName);
      let message = '';
      if (recent.length > 0) {
        // respond with name is at recent[0].short_address
        const person = recent[0];
        message = `${slotName} is ${person.is_intransit === 0 ? 'moving near' : 'at'} ${person.address1}`;
      } else {
        // respond with I couldnt find name
        message = `I could not find ${slotName}`
      }
      response.say(message);
    });

    next();
  },
  function setupApp(next) {
    winston.info("Configuring server");

    // accept JSON
    app.use(bodyparser.json());

    // log all requests to disk
    app.use(expressWinston.logger({
      transports: [
        new winston.transports.File({
          filename: ACCESS_LOG,
          json: false
        })
      ]
    }));

    // webhook event from life360
    app.all("/webhook", (req, res) => {
      winston.info('Webhook call received' + JSON.stringify(req.params) + JSON.stringify(req.headers));
      // if (req.headers["access_token"] === state.access_token) {
      //   return res.status(401).send({ errorMessage: "Invalid access token" }).end();
      // }
      refreshState();
    });



    app.get('/check', (req, res, next) => {
      let output = "Life 360 Plugin";
      output += "<style>h3{margin-bottom:0}</style>"
      output += "<h3>Places</h3><br />"
      output += "<pre><code>";
      output += JSON.stringify(state.places.map(x => new Object({
        name: x.name,
        latitude: x.latitude,
        longitude: x.longitude,
        radius: x.radius
      })), null, 4);
      output += "</code></pre>";
      output += "<h3>Circles</h3><br />"
      output += "<pre><code>";
      output += JSON.stringify(state.circles.map(x => new Object({
        name: x.name
      })), null, 4);
      output += "</code></pre>";
      output += "<h3>Circles</h3><br />"
      output += "<pre><code>";
      output += JSON.stringify(state.members.map(x => formatLocation(x)), null, 4);
      output += "</code></pre>";
      output += "<h3>Alexa Skill Builder</h3><br />";
      output += "<pre><code>";
      output += JSON.stringify(alexaApp.schemas.skillBuilder(), null, 4);
      output += "</code></pre>";
      res.send(output).end();
    });

    // log all errors to disk
    app.use(expressWinston.errorLogger({
      transports: [
        new winston.transports.File({
          filename: ERROR_LOG,
          json: false
        })
      ]
    }));

    // proper error messages with Joi
    app.use((err: any, req: Request, res: Response, next: NextFunction) => {
      winston.error(err);

      res.status(500).send({
        errorMessage: 'An unhandled error occured'
      });
    });

    if (isProd === true && (config.cert_file.length === 0 || config.key_file.length === 0)) {
      next('SSL Certs are required because this addon receives data from a third party so we want their data safely');

    } else if (isProd === false) {
      winston.info('NO certificate files found, listing via http');
      app.listen(PORT, next);
    } else {
      winston.info('Listing via https');
      const options = {
        cert: fs.readFileSync(path.join('/ssl/', config.cert_file)),
        key: fs.readFileSync(path.join('/ssl/', config.key_file))
      };
      https.createServer(options, app as any).listen(PORT, next);

    }

  }
], (error) => {
  if (error) {
    return winston.error(<any>error);
  }
  winston.info("Listening at port %s", PORT);
});

life360/rootfs/server.ts
/**
 * @author Eric Ditter
 * @file Hassio addon that polls Life360 API for location changes as
 *       well as listens via webhook for life360 initiated changes.
 */

import { Circle, Member, Place, PlacesRequest, CirclesRequest, TokenRequest, MembersRequest, Location } from './life360';

import * as winston from 'winston';

import * as mqtt from 'mqtt';
import * as async from 'async';
import * as jsonfile from 'jsonfile';
import * as fs from 'fs';
import * as path from 'path';
import Axios from 'axios';
import { Response, Request, NextFunction, json } from 'express';
import { v4 } from 'uuid';
// import { map } from 'async';
import { MqttClient } from 'mqtt';
const baseAuthKey = 'cFJFcXVnYWJSZXRyZTRFc3RldGhlcnVmcmVQdW1hbUV4dWNyRUh1YzptM2ZydXBSZXRSZXN3ZXJFQ2hBUHJFOTZxYWtFZHI0Vg==';
console.log('Starting...TS File');
const CONFIG_DIR = process.env.CONFIG_DIR || './data',
  STATE_FILE = path.join(CONFIG_DIR, 'state.json'),
  OPTIONS_FILE = path.join(CONFIG_DIR, 'options.json'),
  CURRENT_VERSION: string = jsonfile.readFileSync(path.join('package.json')).version,
  isProd = process.env.IS_PROD === 'true',
  PORT = 8081,
  baseUrl = 'https://api.life360.com/v3',
  config = jsonfile.readFileSync(OPTIONS_FILE) as {
    mqtt_host: string;
    mqtt_port: number;
    mqtt_username: string;
    mqtt_password: string;
    preface: string;
    cert_file: string;
    key_file: string;
    host_url: string;
    life360_user: string;
    life360_password: string;
    refresh_minutes: number;
    process_type: 'MQTT' | 'HTTP',
    user_device_map: { life360_name: string, known_devices_name: string }[]
  },
  state = loadSavedState({
    saved_token: null as string | null,
    circles: [] as { id: string, name: string }[],
    history: {},
    members: [] as Member[],
    places: [] as Place[],
    access_token: v4(),
    CURRENT_VERSION
  });

let client: MqttClient | null = null;
config.process_type = <any>config.process_type.toUpperCase();
if (config.process_type === 'MQTT') {
  const url = config.mqtt_host.includes('://') ? config.mqtt_host : `mqtt://${config.mqtt_host}`;
  client = mqtt.connect(url, {
    username: config.mqtt_username,
    password: config.mqtt_password,
    port: config.mqtt_port,
  });
} else if (config.process_type === 'HTTP') {
  if (!config.user_device_map || config.user_device_map.length === 0) {
    winston.error(`When using HTTP you must have a value in user_device_map`);
    process.exit(1);
  }
} else {
  winston.error(`process_type must be either MQTT or HTTP. ${config.process_type} is invalid`);
  process.exit(1);
}

function isServer() {
  return config.host_url && config.host_url.length > 0 && config.cert_file.length > 0 && config.key_file.length > 0;
}

winston.remove(winston.transports.Console);
winston.add(winston.transports.Console, {
  timestamp: () => {
    const
      date = new Date(),
      month = date.getMonth() + 1,
      day = date.getDate(),
      hour = date.getHours(),
      minute = date.getMinutes(),
      second = date.getSeconds(),
      hourFormatted = hour % 12 || 12,
      minuteFormatted = minute < 10 ? '0' + minute : minute,
      morning = hour < 12 ? 'am' : 'pm';

    return `${month}/${day} ${hourFormatted}:${minuteFormatted}:${second} ${morning}`;
  }
});

function loadSavedState<T>(defaults: T): T {
  try {
    const savedState = jsonfile.readFileSync(STATE_FILE);
    if (savedState.CURRENT_VERSION !== CURRENT_VERSION) {
      return defaults;
    }
    return savedState;
  } catch (ex) {
    winston.info('No previous state found, setting to defaults.');
    return defaults;
  }

}

async function refreshToken() {

  const headers = {
    headers: {
      Authorization: 'basic ' + baseAuthKey
    }
  };
  const authResponse = await Axios.post<TokenRequest>(`${baseUrl}/oauth2/token.json`, {
    grant_type: 'password',
    username: config.life360_user,
    password: config.life360_password
  }, headers);

  if (authResponse.data.access_token) {
    state.saved_token = authResponse.data.token_type + ' ' + authResponse.data.access_token;
  } else {
    throw Error('No token was returned. ' + authResponse.data.errorMessage);
  }

  headers.headers = {
    Authorization: state.saved_token
  };
  const circleResponse = await Axios.get<CirclesRequest>(`${baseUrl}/circles.json`, headers);
  if (Array.isArray(circleResponse.data.circles)) {
    state.circles = [];
    circleResponse.data.circles.forEach(circ => {
      state.circles.push({
        id: circ.id,
        name: circ.name
      });

    });

  } else {
    winston.warn('No circles were returned from the request. ' + circleResponse.data.errorMessage);
  }

  return new Promise<void>((accept, reject) => {
    state.places = [];

    async.forEach(state.circles, async (circ, callback) => {
      const placesResponse = await Axios.get<PlacesRequest>(`${baseUrl}/circles/${circ.id}/places.json`, headers);
      if (Array.isArray(placesResponse.data.places)) {
        placesResponse.data.places.forEach(x => state.places.push(x));
      }

      if (isServer()) {
        await Axios.delete(`${baseUrl}/circles/${circ.id}/webhook.json`, headers);
        const returnUrl = config.host_url + '/webhook?access_token=' + state.access_token;
        const webhookResponse = await Axios.post(`${baseUrl}/circles/${circ.id}/webhook.json`, {
          url: returnUrl
        }, headers);

        if (!webhookResponse.data.hookUrl) {
          winston.warn('Webhook was not created. ' + circleResponse.data.errorMessage);
        }
      } else {
        winston.warn('No Webhook created because url or certs were not set');
      }
      callback();
    }, (err) => {
      if (err) {
        reject(err);

      } else {
        accept();

      }
    });

  });

}

function formatLocation(input: Member) {
  const location = input && input.location ? input.location : null;
  if (!location) {
    return null;
  }
  if (config.process_type === 'MQTT') {

    return {
      longitude: parseFloat(location.longitude),
      latitude: parseFloat(location.latitude),
      gps_accuracy: parseInt(location.accuracy, 0),
      battery_level: parseInt(location.battery, 0),
      name: input.firstName,
      speed: location.speed,
      is_driving: parseInt(location.isDriving, 0),
      is_intransit: parseInt(location.inTransit, 0),
      is_charging: parseInt(location.charge, 0),
      wifi_state: parseInt(location.wifiState, 0),
      address1: location.address1,
      address2: location.address2,
    };

  } else if (config.process_type === 'HTTP') {

    const user = config.user_device_map.filter(x => x.life360_name.toUpperCase() === input.firstName.toUpperCase());
    return {
      dev_id: user.length > 0 ? user[0].known_devices_name : input.firstName,
      gps: [parseFloat(location.latitude), parseFloat(location.longitude)],
      gps_accuracy: parseInt(location.accuracy, 0),
      battery: parseInt(location.battery, 0),
      attributes: {
        speed: location.speed,
        is_driving: parseInt(location.isDriving, 0),
        is_intransit: parseInt(location.inTransit, 0),
        is_charging: parseInt(location.charge, 0),
        wifi_state: parseInt(location.wifiState, 0),
        address1: location.address1,
        address2: location.address2,
      },
      // friendly_name: input.firstName
    };

  }

  throw Error('Invalid process_type');
}

async function refreshState() {
  const topic = `${config.preface}/location`;

  try {
    state.members = await getData();
    const data = state.members.map(x => formatLocation(x));
    state.history = data;
    jsonfile.writeFileSync(STATE_FILE, state, {
      spaces: 2
    });
    data.forEach(msg => {
      if (!msg) {
        return;
      }
      if (config.process_type === 'MQTT') {

        if (client !== null) {
          const userTopic = `${topic}/${msg.name}`;
          winston.info('MQTT Message was written to ' + userTopic);
          client.publish(userTopic, JSON.stringify(msg), {
            qos: 1,
            retain: true
          });
        }
      } else if (config.process_type === 'HTTP') {
        winston.info('HTTP Message sent for ' + msg.dev_id);

        const conf = {
          headers: {

            'x-ha-access': process.env.HASSIO_TOKEN,
            'content-type': 'application/json'
          }
        };

        Axios.post('http://hassio/homeassistant/api/services/device_tracker/see', msg, conf)
          .catch(err => winston.error(err.message));

      }
    });



  } catch (err) {
    winston.error(err.message);
  }
}

async function getData() {
  if (!state.saved_token) {
    await refreshToken();
  }

  return new Promise<Member[]>((accept, reject) => {
    const headers = {
      headers: {
        Authorization: state.saved_token || 'un-auth-request'
      }
    };
    const results: Member[] = [];
    async.forEach(state.circles, (circ, callback) => {
      Axios.get<MembersRequest>(`${baseUrl}/circles/${circ.id}/members.json`, headers)
        .then(membersResponse => {
          membersResponse.data.members.forEach(x => results.push(x));
          callback();
        })
        .catch(err => {
          if (err.status === 401) {
            state.saved_token = null;
          }
          callback(err);
        });
    }, (err) => {
      if (err) {
        reject(err);

      } else {
        accept(results);

      }
    });
  });
}

// main flow
async.series([
  function init(next) {
    refreshToken().then(() => {
      winston.info('init/refresh token successful');
      next();

    }).catch(err => {
      winston.error('Error getting initial token request. Ending process');
      next(err);

    });

  },
  function configurePolling(next) {
    const interval = 60 * 1000 * config.refresh_minutes;
    winston.info(`Configuring auto update for every ${config.refresh_minutes} minutes`);

    // save current state every 15 minutes
    setInterval(refreshState, interval);

    refreshState().then(x => {
      state.circles.forEach((circle, i, arr) => {
        winston.info(`Circle ${i + 1} of ${arr.length}: ${circle.name}`);
      });

      state.members.forEach((member, i, arr) => {
        const memberData = formatLocation(member);
        winston.info(`Member ${i + 1} of ${arr.length}\n Name: ${member.firstName}; Data: ${JSON.stringify(memberData, null, 4)} `);

      });
    });

    process.nextTick(next);
  },
  async function setupApp(next) {
    if (isServer()) {
      winston.info('Configuring server');
      const express = await import('express');
      const https = await import('https');
      // const expressWinston: any = await import('express-winston');
      const bodyparser = await import('body-parser');
      const app = express();

      app.use(bodyparser.json());

      // webhook event from life360
      app.all('/webhook', async (req, res, n) => {
        if (req.query.access_token === state.access_token) {
          winston.info('Webhook call received with valid access token');

          await refreshState();
          return res.send('ok').end();
        } else {
          winston.info('Webhook call was made with invalid access token');
          return res.send('Webhook is visible externally').end();

        }

      });

      app.use((err: any, req: Request, res: Response, n: NextFunction) => {
        winston.error(err.message);

        return res.status(500).send({
          errorMessage: 'An unhandled error occured'
        });
      });
      if (isProd) {
        winston.info('Listing via https');
        const options = {
          cert: fs.readFileSync(path.join('/ssl/', config.cert_file)),
          key: fs.readFileSync(path.join('/ssl/', config.key_file))
        };
        https.createServer(options, app as any).listen(PORT, next);
      } else {
        app.listen(PORT, next);
      }
      winston.info('Listening at port %s', PORT);
    }
    // accept JSON




  }
], (error) => {
  if (error) {
    return winston.error((<any>error).message);
  }
});
